alphadia/workflow/optimization.py

```diff
@@ -579,3 +576,164 @@ def __init__(
+        self.fragment_dfs = []
+
+    def _get_exponential_batches(self, step):
+        """Get the number of batches for a given step
+        This plan has the shape:
+        1, 2, 4, 8, 16, 32, 64, ...
+        """
+        return int(2**step)
+
+    def set_batch_plan(self):
+        """Gets an exponential batch plan based on the batch_size value in the config."""
+        n_eg = self.library._precursor_df["elution_group_idx"].nunique()
+
+        plan = []
+
+        batch_size = self.config["calibration"]["batch_size"]
+        step = 0
+        start_idx = 0
+
+        while start_idx < n_eg:
+            n_batches = self._get_exponential_batches(step)
+            stop_idx = min(start_idx + n_batches * batch_size, n_eg)
+            plan.append((start_idx, stop_idx))
+            step += 1
+            start_idx = stop_idx
+
+        self.batch_plan = plan
+
+    def update_with_extraction(self, feature_df, fragment_df):
+        """Extract features and fragments from the current batch of the optimization lock.
+
+        Parameters
+        ----------
+        feature_df: pd.DataFrame
+            The feature dataframe for the current batch of the optimization lock (from workflow.extract_batch).
+
+        fragment_df: pd.DataFrame
+            The fragment dataframe for the current batch of the optimization lock (from workflow.extract_batch).
+        """
+
+        self.feature_dfs += [feature_df]
+        self.fragment_dfs += [fragment_df]
+
+        self.total_precursors = len(self.features_df)
+
+    def update_with_fdr(self, precursor_df):
+        """Calculates the number of precursors at 1% FDR for the current optimization lock and determines if it is sufficient to perform calibration and optimization.
+
+        Parameters
+        ----------
+        precursor_df: pd.DataFrame
+            The precursor dataframe for the current batch of the optimization lock (from workflow.perform_fdr).
+        """
+
+        self.count = len(precursor_df[precursor_df["qval"] < 0.01])
+
+        self.has_target_num_precursors = self.count >= self.target_count
+
+    def increase_batch_idx(self):
+        """If the optimization lock does not contain enough precursors at 1% FDR, the optimization lock proceeds to include the next step in the batch plan in the library attribute.
+        This is done by incrementing self.batch_idx.
+        """
+        self.batch_idx += 1
+
+    def decrease_batch_idx(self):
+        """If the optimization lock contains enough precursors at 1% FDR, checks if enough precursors can be obtained using a smaller library and updates the library attribute accordingly.
+        If not, the same library is used as before.
+        This is done by taking the smallest step in the batch plan which gives more precursors than the target number of precursors.
+        """
+
+        batch_plan_diff = np.array(
+            [
+                stop_at_given_idx - self.stop_idx * self.target_count / self.count
```

Comment:
Yes, it normally is a non-integer, but this doesn't pose a problem since this is just a way to find the smallest positive difference (after which we then take the index of this position).